app/ts/simulation/priceEstimator.ts

```diff
@@ -21,99 +20,92 @@ export class PriceEstimator {
-					chainId: chainId,
-					nonce: transactionCount + 1n,
-					maxFeePerGas: 0n,
-					maxPriorityFeePerGas: 0n,
-					gas: 15000000n,
-					to: token.token,
-					value: 0n,
-					input: stringToUint8Array(swapInterface.encodeFunctionData('approve', [addressString(UNISWAP_V2_ROUTER_ADDRESS), 2n ** 127n] )),
-					accessList: [],
-				},
-				{
-					type: '1559' as const,
-					from: sender,
-					chainId: chainId,
-					nonce: transactionCount + 2n,
-					maxFeePerGas: 0n,
-					maxPriorityFeePerGas: 0n,
-					gas: 15000000n,
-					to: UNISWAP_V2_ROUTER_ADDRESS,
-					value: 0n,
-					input: stringToUint8Array(swapInterface.encodeFunctionData(
-						'swapTokensForExactETH',
-						[10n ** 18n / 2n, 2n ** 127n, [addressString(token.token), addressString(CHAINS[chainString].weth)], addressString(sender), deadline]
-					)),
-					accessList: [],
-				},
-			]
-			const results = await this.ethereum.multicall(swapTransactions, block.number + 1n)
-			if (results.length !== 3) throw ('invalid multicall result')
-			if (results[2].statusCode === 'success') {
-				const parsed = swapInterface.decodeFunctionResult('swapTokensForExactETH', results[2].returnValue)
-				const inOut = parsed.toObject() as { amounts: bigint[] } // TODO, change to funtype
-				if (inOut.amounts.length != 2) return []
-				if(inOut.amounts[0] <= 0n || inOut.amounts[1] <= 0n || token.decimals <= 0n) return []
-				inOutResults.push( {
-					token: addressString(token.token ),
-					inOutAmount: [inOut.amounts[1], inOut.amounts[0]],
-					decimals: token.decimals
-				})
+			const poolAddresses = calculateUniswapPools(token.address, CHAINS[chainString].weth)
+			const uniswapSpotCalls = getUniswapSpotCalls(token.address, quoteToken.address, poolAddresses)
+
+			const callData = stringToUint8Array(IMulticall3.encodeFunctionData('aggregate3', [uniswapSpotCalls]))
+			const callTransaction = {
+				type: '1559',
+				to: MULTICALL3,
+				value: 0n,
+				input: callData,
 			}
+			const multicallReturnData: { success: boolean, returnData: string }[] = IMulticall3.decodeFunctionResult('aggregate3', await this.ethereum.call(callTransaction, 'latest'))[0]
+
+			const prices = multicallReturnData.reduce((prices, { success, returnData }, index): { price: bigint, k: bigint }[] => {
+				if (index > 4) return prices
+
+				// V2 Pool
+				if (index === 0 && success) {
+					const { reserve0, reserve1 } = IUniswapV2Pool.decodeFunctionResult('getReserves', returnData)
+
+					const price = poolAddresses.token0IsQuote
+						? (reserve1 * (10n ** BigInt(token.decimals))) / reserve0
+						: (reserve0 * (10n ** BigInt(token.decimals))) / reserve1
+
+					return [...prices, { price, k: BigInt(reserve0 * reserve1) }]
+				}
+
+				// V3 Pool
+				if (success && multicallReturnData[index * 2 + 3].success && multicallReturnData[index * 2 + 4].success) {
+					const { sqrtPriceX96 } = IUniswapV3Pool.decodeFunctionResult('slot0', returnData)
+
+					const price = poolAddresses.token0IsQuote
```

Comment:
could this logic be wrapped in a function with a refence to a source where it comes from?

Reply:
the decimals also should already be bigints